### DISCUSSION TITLE
Evaluate Module vs Run Script?

### DISCUSSION THREAD

**Discussion by linrongbin16** [denoland/deno#25880] [Q&A] [answered]
Hi, I'm learning the `deno_core` and the V8 engine inside. I'm confused about the difference between evaluate module and run script.

For example, in the (very simple) [Hello, runjs!](https://deno.com/blog/roll-your-own-javascript-runtime#hello-runjs) sample code in [Roll your own JavaScript runtime](https://deno.com/blog/roll-your-own-javascript-runtime) blog.

Roughly speaking (please correct me if I'm wrong), `deno_core` is doing following steps:

1. `load_main_es_module` uses the [v8::script_compiler::compile_module](https://docs.rs/v8/latest/v8/script_compiler/fn.compile_module.html) to compile the main module, and also created all the module dependencies from it (like a dependency tree). At this time, all the modules that stores on local file system is synchronizely loaded, all the remote modules that need to be downloaded is still in fetching status.
2. `mod_evaluate` uses the [v8::Module::evaluate](https://docs.rs/v8/latest/v8/struct.Module.html#method.evaluate) to evalute the main module.
3. `run_event_loop` start to running the event loop, so all the remote modules can be downloaded, all the `async` and `Promise` can be resolved, until the javascript code in the main module been all executed to done.

But, I'm confused that, there's also an API [v8::Script::run](https://docs.rs/v8/latest/v8/struct.Script.html#method.run). Can I simply create a v8 `Script` and then run it, instead of load ES module and evaluate it?

What's the difference between these two groups of APIs?

1. The module can only be evaluated once, the script can be run multiple times (I think I read this from somewhere).
2. Also here's a guess: the javascript code in a v8 `Script` is pure tranditional js, it cannot use `require` or `import` keywords to import other modules. Correct?

**Comment by marvinhagemeister** [comment]
This page has more information about the differences between scripts and modules in JavaScript engines https://v8.dev/features/modules

**Comment by bartlomieju** [comment, marked as answer]
> load_main_es_module uses the [v8::script_compiler::compile_module](https://docs.rs/v8/latest/v8/script_compiler/fn.compile_module.html) to compile the main module, and also created all the module dependencies from it (like a dependency tree). At this time, all the modules that stores on local file system is synchronizely loaded, all the remote modules that need to be downloaded is still in fetching status.

By the time `load_main_es_module` all remote modules need to be downloaded and loaded into the V8 as well

> run_event_loop start to running the event loop, so all the remote modules can be downloaded, all the async and Promise can be resolved, until the javascript code in the main module been all executed to done.

Remote modules are already downloaded, this function keeps going as long as there are async actions pending (think timers, reads from net socket, etc).

> Also here's a guess: the javascript code in a v8 Script is pure tranditional js, it cannot use require or import keywords to import other modules. Correct?

Correct

**Reply by linrongbin16** [reply]
Really thanks for your knowledge sharing and teaching!
